### PR TITLE
update: regexp, fix: displayMode: true in block level, fix: changed katex css version to latest

### PR DESCRIPTION
**Pull request by Rafi-Riday**
Hei, I changed the regexp, the code works a little different now, you can see here : https://marked-katex-extension.netlify.app/
The current version does not supports function like `$\fcolorbox{red}{aqua}{$\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}$}$` or like this:
```latex
$$
\fcolorbox{red}{aqua}{$\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}$}
$$
```
In my version all of those works fine, but a little bit update has to be done. And also, the current tests are failing because of the newer rules, I think you should make new tests if you think thease updates are better. Then we have to check edge cases.
Any `$...$` inside another `$...$` works now if in 1 line, like the previous codes, but they dont work if are splitted into multiple lines. We may update it to work also in multiple lines. 

I changed the `displayMode: true` for block level output which is the proper way to render block level cause katex has some environment which does not works without that : https://katex.org/docs/supported.html#environments

I changed katex css version to latest in `README` and add a simple instruction to use `fleqn: true` to make the block level katex output left aligned.

### CODE REVIEW

**Comment by UziTech** [comment]
Sounds good. I don't think we need to mention `fleqn` in the readme we can just link to all of the katex options and they can read what they do there.

If you can get tests to pass with new tests we can merge this.

**Comment by Rafi-Riday** [comment]
All right, linking all katex options is great, but the problem is, I cant add more tests here, I dont know how to do it in this case🙁. Can you do the tests and show me what is the problems? Also I made an website where you can test for all the cases : https://marked-katex-extension.netlify.app/
If you think my changes will pass all the use cases, you may merge it, then update the test scripts, or you may directly give me the test scripts to include them in my version. I tried all normal use cases, just some more functionality to add, like the proper wrking `\$` which currently does not works 100% properly, like `$\pi\$$` will not work, but `$\$\pi$` will work fine, so just a little updates needed to make the regex better, I think strict rules like only 1 `$` for inline and only 2 `$$` for block katex will make it easy to introduce those updates, where it will not match the `\$` as a katex input but will match the outside wrapper.
Hope you will test my updates, I'm new in testing😞

**Comment by UziTech** [comment]
You can look at #167 to see how I added tests. You can just add tests to the snapshots object in index.test.js

https://github.com/UziTech/marked-katex-extension/blob/87620ef9a1bf1203d13db7ac28a16589a7e35315/spec/index.test.js#L9

The key is the test name the value is the test markdown.

When you run `npm t` jest will create snapshots for your new tests in https://github.com/UziTech/marked-katex-extension/blob/main/spec/__snapshots__/index.test.js.snap

You can look in that file to see if the snapshot is correct. You can recreate snapshots with `npm t -- -u`

**Comment by Rafi-Riday** [comment]
Oh I get it now, thanks for the instructions, i added more tests. including `$\n...\n$`, but do you think we should remove this feature? or this can stay?

I added more tests and the tests ran successfully. I ran all the commands, also fixed lint error. So maybe now you can merge, but we have to make some more changes.

**Comment by Rafi-Riday** [comment]
hei, should we add details of this extension in the README.md? Or should we create another demo website where others can check it out? Well I already have a demo checking website : https://marked-katex-extension.netlify.app/
If you say I can add detailed documentation in this website, and we may need to update README then. I think adding a demo website would be great

**Comment by Rafi-Riday** [comment]
So I get another bug, we cant use multiple lines inside block katex like this :
```latex
$$
\begin{smallmatrix}
   a & b \\
   c & d
\end{smallmatrix}
$$
```
Here multiple line is not working, and the bug is from the current extension, you can check here : https://svelte.dev/repl/cc23acc2478a4d97bc88958a4d6b30e3?version=4.0.4 where I'm using current version of marked-katex-extension.

But the best thing is, its working when I'm writing the whole katex code in 1 line like this :
```latex
$$
\begin{smallmatrix} a & b \\ c & d \end{smallmatrix}
$$
```
So we have to make multiline katex code work asap. I'm on it :

File: README.md [hunk 1]
```diff
@@ -39,6 +44,8 @@ marked.use(markedKatex(options));
 marked.parse("katex: $c = \\pm\\sqrt{a^2 + b^2}$");
 ```
 
+# Use `fleqn: true` in `options` to align block level katex output to the left
```
[the remark this answers is not in the document]
Comment (Rafi-Riday) [reply README.md]: Are you talking about changing the README? But I already updated it😐
https://github.com/UziTech/marked-katex-extension/pull/171/commits/2b225a668845dbc68b911663c1562c9fad70c89c
I changed this file after you ran the tests here

File: src/index.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ function inlineKatex(options) {
     level: 'inline',
     start(src) { return src.indexOf('$'); },
     tokenizer(src, tokens) {
-      const match = src.match(/^\$+([^$\n]+?)\$+/);
+      const match = src.match(/^\$+((?:\n)?\S(?:.*\S)?(?:\n)?)\$+/);
```
[the remark this answers is not in the document]
Comment (Rafi-Riday) [reply src/index.js:18]: Actually there can be many opinions here. But if you see basic rules for usng markdown, like for bold you write `**bold**`, but if you  write `**bold **` or `** bold**` it wont work, only `**...**` works, now there are some functions which works with spaces also, like the inline code \`. but it has its own problem, for suppose, in the line below, just putting a backtick in wrong place will render different result :
abcd `code` ef`gh `code` ijkl

Here a backtick in the middle of efgh will render the output very differently, here i wanted to generate inline code only the `code` strings. Well we can escape the backtick with \\ , but currently we cant escape `$` with backtick like `\$`. We dont have that feature right now, not in the current extension version, not in my version, my version does not support \\$ fully, the current extension version supports \\$ outside of katex. but katex has its own syntax `\$` in order to render `$`, in my version, this support is partial, but `$\$$` does not work. This same code also wont work in current extension. But in my version, `$abcd\$efgh$` works fine, so basically its partial support, instead of just throwing error. later we may add a different functionality for `\$` so I think we should not allow space inside `$...$`. Well if we do allow, I doont know the `\$` inside katex does not work. Just like in the current version, `\$` is not working.

Also, syntax like `**...**` is more better than syntax like code (`), I'm using katex for a while, I think it as a better way, also some editors do the same. for example stackedit.io, which also doesn't support `\$` inside katex. But stricting `$...$` and not allowing space is not a problem I think, it just makes it better to detect `$...$`. But I think both way is enough to work with, now you decide and tell me what  should we do, then I will update the code 💯

**Comment by Rafi-Riday** [comment]
# Wait, let me fix some bugs I fount, dont merge it now

File: src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (UziTech) [reply src/index.js:18]: Markdown is meant to be easy to read for bold it makes sense that `this is **bold**` is easier to read than `this is ** bold **` with bold it is easier to understand what is meant when there are no spaces. For katex I think `$ c = \\pm\\sqrt{a^2 + b^2} $` is easier to read than `$c = \\pm\\sqrt{a^2 + b^2}$` that `$c` can be easily confused as one variable at a glance.

I think to prevent something like `$300 $400` being seen as katex we should require a space outside the `$...$` so `This is katex:$ c = \\pm\\sqrt{a^2 + b^2} $` does not count as katex there will have to be a space between `: $` so it is easier to read.

**Review by UziTech — CHANGES_REQUESTED** [review]
please remove pnpm-lock.yaml

File: src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (UziTech) [reply src/index.js:18]: also anything that parses text with newlines should be in a block level tokenizer. We don't want to allow newlines with other content like 

```
this is katex: $
c = \\pm\\sqrt{a^2 + b^2}
$
```

We should only allow newlines when the first `$` is at the beginning of the line like

```
this is katex:
$
c = \\pm\\sqrt{a^2 + b^2}
$
```

[the remark this answers is not in the document]
Comment (Rafi-Riday) [reply src/index.js:18]: > Markdown is meant to be easy to read for bold it makes sense that `this is **bold**` is easier to read than `this is ** bold **` with bold it is easier to understand what is meant when there are no spaces. For katex I think `$ c = \\pm\\sqrt{a^2 + b^2} $` is easier to read than `$c = \\pm\\sqrt{a^2 + b^2}$` that `$c` can be easily confused as one variable at a glance.
> 
> I think to prevent something like `$300 $400` being seen as katex we should require a space outside the `$...$` so `This is katex:$ c = \\pm\\sqrt{a^2 + b^2} $` does not count as katex there will have to be a space between `: $` so it is easier to read.

Well we can use `\$300 \$400` to prevent silly issue like that, also we can allow space, no problem here, the current extension is actually ok to work with, I agree, but some features are not being supported, like `$\fcolorbox{red}{aqua}{$\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}$}$` or this :
```tex
$$
\fcolorbox{red}{aqua}{$\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}$}
$$
```
I think my version of the block level is ok, its supporting the codes, but the current inline version does not support the same code inline because of `$...$....$...$` and in my vversion, this is matched, so in my version its a bug, the regexp cant detect 2 different katex, rather matches as a same katex, so `katex : $\pi$ and $\theta$` will not work in my version, it will match `\pi$ and $\theta`. in many cases there are multiple inline katex in single line, but block katex has to be in different line, so I guess we need to get extra thing here, like we can introduce another matching regex, for something like \`$...$\`, in order to allow a katex inside another katex, and if the case is not, then normal katex like `$...$`, but the user will have both functionality to work with. How abouth this idea? And also in this version of extension the `\$` does not work inside katex, how do we add this support? `\$` is valid katex code.

**Comment by UziTech** [comment]
I think #167 is ready to merge. It can handle `$\pi$ and $\theta$` as well as `\$`. For `\$` there does need to be a space after it (`$\$ $` works but `$\$$` does not work). It sets `displayMode = true` on `$$` for inline and block level katex.

**Comment by Rafi-Riday** [comment]
> I think #167 is ready to merge. It can handle `$\pi$ and $\theta$` as well as `\$`. For `\$` there does need to be a space after it (`$\$ $` works but `$\$$` does not work). It sets `displayMode = true` on `$$` for inline and block level katex.

All right, you keep going, I'll do some more research about this, then I'll come back again
